src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePal.cs

```diff
@@ -19,82 +19,160 @@ public static ICertificatePal FromHandle(IntPtr handle)
             return new OpenSslX509CertificateReader(Interop.libcrypto.X509_dup(handle));
         }
 
-        public static unsafe ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
+        public static ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
         {
-            // If we can see a hyphen, assume it's PEM.  Otherwise try DER-X509, then fall back to DER-PKCS12.
-            SafeX509Handle cert;
+            ICertificatePal cert;
 
-            // PEM
-            if (rawData[0] == '-')
+            if (TryReadX509Der(rawData, out cert) ||
+                TryReadX509Pem(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs7Der(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs7Pem(rawData, out cert) ||
+                PkcsFormatReader.TryReadPkcs12(rawData, password, out cert))
             {
-                using (SafeBioHandle bio = Interop.libcrypto.BIO_new(Interop.libcrypto.BIO_s_mem()))
+                if (cert == null)
                 {
-                    Interop.libcrypto.CheckValidOpenSslHandle(bio);
-
-                    Interop.libcrypto.BIO_write(bio, rawData, rawData.Length);
-                    cert = Interop.libcrypto.PEM_read_bio_X509_AUX(bio, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
+                    // Empty collection, most likely.
+                    throw new CryptographicException();
```

Comment:
Should we have a descriptive message for this case?  Something about an unsupported format?